File: cp_rust/src/bin/cses-3399-raab-game-i.rs
// Created by Ayush Biswas at 2025/07/11 17:09
// https://cses.fi/problemset/task/3399
use cp_lib::*;

// @code begin
use cpio::*;

sol_n! {
    fn solution(
        [n, a, b]: [usize; 3]
    ) -> CPResult<Lines<String>, String> {
        if a + b > n {
            return Err("NO".into()).into()
        }
        if a * b == 0 && a + b != 0 {
            return Err("NO".into()).into()
        }
        let m = a + b;
        Ok(vec![
            "YES".into(),
            (1..=n).collect::<Words<_>>().to_string(),
            (m - b + 1..=m).chain(1..=m-b).chain(m+1..=n).collect::<Words<_>>().to_string()
        ].into()).into()
    }
}

// @code end

File: cp_haskell/bin/cses-3399-raab-game-i.hs
-- Created by Ayush Biswas at 2026/03/08 14:24
-- https://cses.fi/problemset/task/3399

-- @code begin

import Control.Monad (replicateM)
import Text.Printf (printf)

raab :: [Int] -> IO ()
raab [n, a, b]
  | (a == 0 || b == 0) && (m /= 0) = putStrLn "NO"
  | a + b == 1 = putStrLn "NO"
  | a + b > n = putStrLn "NO"
  | otherwise = do
      putStrLn "YES"
      mapM_ (printf "%d ") [1 .. n]
      printf "\n"
      mapM_ (printf "%d ") ([a + 1 .. m] ++ [1 .. a] ++ [m + 1 .. n])
      printf "\n"
  where
    m = a + b

getInts :: IO [Int]
getInts = map read . words <$> getLine

main :: IO ()
main = do
  t :: Int <- read <$> getLine
  nabs :: [[Int]] <- replicateM t getInts
  mapM_ raab nabs

-- @code end